iwyu_include_picker.cc

```diff
@@ -1329,13 +1329,32 @@ vector<string> IncludePicker::GetCandidateHeadersForFilepathIncludedFrom(
   // ConvertToQuotedInclude because it avoids trouble when the same
   // file is accessible via different include search-paths, or is
   // accessed via a symlink.
-  const string& quoted_include_as_typed
-      = MaybeGetIncludeNameAsWritten(including_filepath, included_filepath);
-  if (!quoted_include_as_typed.empty()) {
-    vector<string>::iterator it = std::find(retval.begin(), retval.end(),
-                                            quoted_includee);
-    if (it != retval.end())
-      *it = quoted_include_as_typed;
+  vector<string>::iterator it = std::find(retval.begin(), retval.end(),
+                                          quoted_includee);
+  if (it != retval.end()) {
+      const string& quoted_include_as_typed
+          = MaybeGetIncludeNameAsWritten(including_filepath, included_filepath);
+      if (!quoted_include_as_typed.empty()) {
+          *it = quoted_include_as_typed;
+      } else {
+          string strippedIncludee = included_filepath;
+          // Directory of the file is implicitly in the search path.
+          // So if we have the full path, shorten it
+          const string implicitHeaderSearchPath =
+                  CanonicalizeHeaderSearchPath(GetParentPath(including_filepath));
+          if (StripLeft(&strippedIncludee, implicitHeaderSearchPath)) {
+              VERRS(6) << "Adding include in same directory for "
+                       << included_filepath << ": " << strippedIncludee
+                       << "\n";
+              // If the included path is absolute, we prefer the short one
+              // else we prefer the current one, which is probably relative to
+              // a search directory
+              if(IsAbsolutePath(StripQuotes(*it)))
+                  retval.insert(it, "\"" + strippedIncludee + "\"");
+              else
+                  retval.push_back("\"" + strippedIncludee + "\"");
+          }
+      }
```

Comment:
OK, accidental erroneous absolute path comes from `ConvertToQuotedInclude(including_filepath, MakeAbsolutePath(GetParentPath(...` `GetParentPath` uses [`llvm::sys::path::parent_path`](http://llvm.org/docs/doxygen/html/namespacellvm_1_1sys_1_1path.html#a734605aeef896f419c79c53d8af8f2e2) and it doesn't add trailing slash. In `GetParentPath` I tried to ensure non-empty path always has trailing slash and it didn't break anything. Moreover, I reverted this block I'm commenting on in PR back to

```c++
  const string& quoted_include_as_typed
      = MaybeGetIncludeNameAsWritten(including_filepath, included_filepath);
  if (!quoted_include_as_typed.empty()) {
    vector<string>::iterator it = std::find(retval.begin(), retval.end(),
                                            quoted_includee);
    if (it != retval.end())
      *it = quoted_include_as_typed;
  }
```
and new tests are still passing. Can you please verify trailing slash helps? Maybe there are other cases where your change is required but these cases aren't tested. Then we need better test coverage.